File: android/app/src/main/kotlin/com/example/kimoi/extractors/Mp4uploadExtractor.kt
package com.example.kimoi.extractors
import com.example.kimoi.model.Video
import dev.datlag.jsunpacker.JsUnpacker
import okhttp3.Headers
import org.jsoup.nodes.Document
import java.net.HttpURLConnection
import java.net.URL

class Mp4uploadExtractor() {

    fun videosFromUrl(url: String, headers: Headers, prefix: String = "", suffix: String = ""): List<Video> {
        val newHeaders = headers.newBuilder()
            .add("referer", REFERER)
            .build()


        val myURL = URL(url)
        val connection = myURL.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.setRequestProperty("Accept", "application/json")
        connection.setRequestProperty("referer", "https://mp4upload.com/")
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36")
        connection.connect()

        val inputStream = connection.getInputStream()
        val response = inputStream.bufferedReader().readText()

        val doc = Document(url).parser().parseInput(response, url)

        val script = doc.selectFirst("script:containsData(eval):containsData(p,a,c,k,e,d)")?.data()
            ?.let(JsUnpacker::unpackAndCombine)
            ?: doc.selectFirst("script:containsData(player.src)")?.data()
            ?: return emptyList()

        // Close the input stream.
        inputStream.close()

        val videoUrl = script.substringAfter(".src(").substringBefore(")")
            .substringAfter("src:").substringAfter('"').substringBefore('"')

        val resolution = QUALITY_REGEX.find(script)?.groupValues?.let { "${it[1]}p" } ?: "Unknown resolution"
        val quality = "${prefix}Mp4Upload - $resolution$suffix"

        return listOf(Video(videoUrl, quality, videoUrl,"Mp4Upload" ,headers =  newHeaders))
    }

    companion object {
        private val QUALITY_REGEX by lazy { """\WHEIGHT=(\d+)""".toRegex() }
        private const val REFERER = "https://mp4upload.com/"
    }
}

File: lib/src/utils/extractors/mp4upload_extractor.dart
import 'package:flutter/material.dart';
import 'package:flutter_inappwebview/flutter_inappwebview.dart';
import 'package:http/http.dart';
import 'package:kimoi/src/infrastructure/infrastructure.dart';
import 'package:kimoi/src/utils/extensions/extension.dart';

class Mp4UploadExtractor {
  final client = Client();

  final headers = {
    "referer": "https://mp4upload.com/",
    "User-Agent":
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36"
  };

  RegExp qualityRegex = RegExp(r"""\WHEIGHT=(\d+)""");

  Future<List<Video>> videosFromUrl(String url) async {
    String? html;

    bool isLoading = true;

    try {
      HeadlessInAppWebView(
        initialUrlRequest: URLRequest(url: Uri.parse(url), headers: headers),
        onLoadStop: (controller, url) async {
          html = await controller.getHtml();
          isLoading = false;
        },
      )
        ..run()
        ..dispose();

      // Future.delayed(const Duration(seconds: 7), () {
      //   isLoading = false;
      // });

      while (isLoading) {
        await Future.delayed(const Duration(milliseconds: 500));
      }

      final script = html!.substringAfter('player.src(');

      final videoUrl = script
          .substringBefore(")")
          .substringAfter("src:")
          .substringAfter('"')
          .substringBefore('"');


      final resolution = qualityRegex.firstMatch(script)?.group(1);

      print(resolution);

      final quality = 'Mp4Upload - ${resolution}p';

      print(resolution);

      if (resolution == null || videoUrl.isEmpty) return [];

      return [
        Video(
            url: videoUrl,
            quality: quality,
            videoUrl: videoUrl,
            headers: headers)
      ];
    } on Exception catch (e) {
      debugPrint('$e');
      return [];
    }
  }
}
